### PR TITLE
PR Webhook

### PR DESCRIPTION
**Pull request by StuartHarris**
- [x] listen for PR webhook using https://github.com/go-playground/webhooks for now
Should probably switch to https://github.com/google/go-github
- [x] checkout PR into an in-memory filesystem
- [x] cat yaml files
- [x] apply on cluster

### CODE REVIEW

File: github/changedProjects.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	}
+	diff, err := headTree.Diff(baseTree)
+	for _, change := range diff {
+		name := change.To.Name
+		if name == "" {
+			name = change.From.Name
+		}
+		dir := getTopLevelDirName(name)
+		if dir != "" {
+			files = appendIfMissing(files, dir)
```
Comment (charypar) [inline github/changedProjects.go]: Not a biggie, but would it be nicer to collect the names into a set? You can use a map for that kind of thing:

```
set := make(map[int]bool)
```

Then just loop over it to get the slice of unique names.

File: deploy.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	if err != nil {
+		log.Fatalf("Error parsing api URL %v\n", err)
+	}
+	if baseEndpoint.Hostname() == "api.github.com" {
+		baseEndpoint.Path = ""
+	} else {
+		baseEndpoint.Path = "/api/v3"
+	}
+
+	fs, changedDirs, err := gh.GetRepo(
```
Comment (charypar) [inline deploy.go]: Would it be nicer to pull the two operations apart? Getting the repo, then doing the diff

File: github/getRepo.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	r, err := git.CloneContext(context, memory.NewStorage(), fs, &git.CloneOptions{
+		URL:  url,
+		Auth: &gHttp.BasicAuth{Username: "none", Password: token},
+	})
+	if err != nil {
+		err = fmt.Errorf("Cannot clone github repo (%s): %v", url, err)
+		return
+	}
+
+	changedDirs, err = GetChangedProjects(r, headRef, baseRef)
```
Comment (charypar) [inline github/getRepo.go]: This feels like it doesn't really belong here - too many jobs at once.

**Review by charypar — CHANGES_REQUESTED** [review]
Nothing major, some minor improvement suggestions.

File: github/changedProjects.go (hunk 1 above)
Comment (StuartHarris) [reply github/changedProjects.go]: Totally get your point but I like the symmetry with `append(slice, element)` and as we're working with slices anyway, I think this is simpler and more intuitive. The map may well perform better, but the function does exit early and we're only ever going to be looping round the changed top level dirs in the repo, so not big numbers of elements.

File: github/getRepo.go (hunk 3 above)
Comment (StuartHarris) [reply github/getRepo.go]: Great point, will refactor

File: github/changedProjects.go (hunk 1 above)
Comment (StuartHarris) [reply github/changedProjects.go]: @charypar thanks - have refactored.

File: github/changedDirectories.go [hunk 4]
```diff
@@ -62,7 +62,7 @@ func GetChangedProjects(repo *git.Repository, headRef, baseRef string) (files []
 		}
 		dir := getTopLevelDirName(name)
 		if dir != "" {
-			files = appendIfMissing(files, dir)
+			directories = appendIfMissing(directories, dir)
```
Comment (charypar) [inline github/changedDirectories.go]: Still not a fan of this. Working with unique strings is just asking to be a set, not a list (slice). Not a deal breaker though.

**Review by charypar — APPROVED** [review]

**Pull request merged**